File: trab3/ruby/newtonraphson.rb
# encoding: utf-8
# Algoritmo NEWTON-RAPHSON
require "./base"

Calcnum::algoritmo :newtonraphson do |x0|
  def df(x) (f(x + tol) - f(x - tol)) / (2 * tol); end
  iterar do
    x = x0 - f(x0) / df(x0)
    break x if (x - x0).abs < tol and f(x).abs < tol
    x0 = x
  end
end

File: trab3/python/newtonraphson.py
from base import *


@algoritmo
def newtonraphson(self, f, x0):
    df = lambda x: (f(x + self.tol) - f(x - self.tol)) / (2 * self.tol)
    self.tabela('k', 'x0', 'f(x0)', 'f\'(x0)', 'x')
    for k in xrange(self.nmi):
        fx0, dfx0 = f(x0), df(x0)
        x = x0 - fx0 / dfx0
        self.tabela(k, x0, fx0, dfx0, x)
        if abs(x - x0) < self.tol and abs(f(x)) < self.tol:
            return x
        x0 = x
    raise ErroNMI

